spec/services/bulk_invoice_service_spec.rb

```diff
@@ -11,6 +11,17 @@
 
       expect(Delayed::Job.last.payload_object.method_name).to eq :start_pdf_job_without_delay
     end
+
+    it "creates a PDF invoice" do
+      order = create(:completed_order_with_fees)
+      order.bill_address = order.ship_address
+      order.save!
+
+      service.start_pdf_job([order.id])
+      Delayed::Job.last.invoke_job
+
+      expect(service.invoice_created?(service.id)).to be_truthy
+    end
```

Comment:
Doesn't it make more sense to have such integration-like spec in the controller and stick to unit-level in this spec file?

Reply:
I'm testing only the interface of this service here. What makes it integration-like? Do you mean the creation of an order in the database?

How would you unit-test the logic of this method?

Reply:
Good point @sauloperez. This test isn't about Delayed Job at all. Your comment made me realise that I can invoke the method without delayed job to test it's actual execution. Have a look at my new commit. That's much clearer.

Also big thanks for the Sandy Metz talk about unit testing. Very simple and clear. My conclusion is though that I don't want to introduce doubles in this spec for several reasons:

- Doubles need to be kept in sync with the API of the thing they are imitating. Sandy mentions that you should have tools for that in order to make this stable and efficient. I don't think we have that in place. We don't have a double which verifies that the doubled method exists on the real object.
- This spec currently uses the public API of the service. Mocking requires knowledge of the internal workings. And that is bad, because changing the implementation of the service then requires changing the spec as well. The current spec should not need adjustments as long as the API stays the same.
- Sandy's main reason for doubles is to stay close to the object under test. We don't want to call `create_pdf` and then check for a file on the filesystem within the spec. The filesystem is far away. But my spec doesn't go far away at all. It uses only the API of the object under test.
- The remaining valid reason for doubles is speed. In this case though, I think that all the stability gained is worth the added 3 seconds.

What do you think?

Reply:
> We don't have a double which verifies that the doubled method exists on the real object.

That talk is old and RSpec's `instance_double` didn't exist at that time. Fortunately, things are better now and we already use it in our tests :tada: !

Reply:
@sauloperez 
> I'm not a big fan of this handle_asynchronously magic

I totally agree! We should avoid that in future code.

> That talk is old and RSpec's instance_double and class_double didn't exist at that time. 

Thank you! I didn't know that. I definitely want to use that more. (Just not here. ;-) )

The outcome feels good now. :+1: